androguard/core/bytecodes/apk.py

```diff
@@ -804,39 +796,63 @@ def _format_value(self, value):
         return value
 
-    def get_element(self, tag_name, attribute, **attribute_filter):
-        """
-        Return element in xml files which match with the tag name and the specific attribute
-
-        :param tag_name: specify the tag name
-        :type tag_name: string
-        :param attribute: specify the attribute
-        :type attribute: string
-
-        :rtype: string
-        """
-        for i in self.xml:
-            if self.xml[i] is None:
-                continue
-            tag = self.xml[i].findall('.//' + tag_name)
-            if len(tag) == 0:
-                return None
-            for item in tag:
-                skip_this_item = False
-                for attr, val in list(attribute_filter.items()):
-                    attr_val = item.get(self._ns(attr))
-                    if attr_val != val:
-                        skip_this_item = True
-                        break
-
-                if skip_this_item:
-                    continue
-
-                value = item.get(self._ns(attribute))
-
-                if value is not None:
-                    return value
-        return None
+    def get_all_attribute_value(
+        self, tag_name, attribute, **attribute_filter
+    ):
+        tags = self.find_tags(tag_name, **attribute_filter)
+        for tag in tags:
+            value = tag.get(attribute) or tag.get(self._ns(attribute))
+            if value is not None:
+                yield self._format_value(value)
+
+    def get_attribute_value(
+        self, tag_name, attribute, **attribute_filter
+    ):
+
+        for value in self.get_all_attribute_value(
+                tag_name, attribute, **attribute_filter):
+            if value is not None:
+                return value
+
+    def get_value_from_tag(self, tag, attribute):
+        return tag.get(self._ns(attribute)) or tag.get(attribute)
+
+    def find_tags(self, tag_name, **attribute_filter):
+        all_tags = [
+            self.find_tags_from_xml(
+                i, tag_name, **attribute_filter
+            )
+            for i in self.xml
+        ]
+        return [tag for tag_list in all_tags for tag in tag_list]
+
+    def find_tags_from_xml(
+        self, xml_name, tag_name, **attribute_filter
+    ):
+        xml = self.xml[xml_name]
+        if xml is None:
+            return []
+        if xml.tag == tag_name:
+            if self.is_tag_matched(
+                xml.tag, **attribute_filter
+            ):
+                return [xml]
+            return []
+        tags = xml.findall(".//" + tag_name)
+        return [
+            tag for tag in tags if self.is_tag_matched(
+                tag, **attribute_filter
+            )
+        ]
+
+    def is_tag_matched(self, tag, **attribute_filter):
+        if len(attribute_filter) <= 0:
+            return True
+        for attr, value in attribute_filter.items():
+            _value = tag.get(attr) or tag.get(self._ns(attr))
```

Comment:
so this tackles the problem of no namespace, right?
Would it be a good idea to add some warning message if the tag should have a namespace but no one is found? But the question is how to know for each attribute if a namespace would be required...
The next question is what to do with files that do stuff like:
`<uses-permission android:name="some dangerous permission" name="some normal permission" />`
these would trick androguard to read the normal permission and not the dangerous one.

Reply:
How would android os parse it? will it take `android:name` or `name` first? We need to test that.

Reply:
depending on that we can either place the namespace based attribute on the first half of `or` statement